common/quotas/health_request_rate_limiter_impl.go

```diff
@@ -0,0 +1,116 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package quotas
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"go.temporal.io/server/common/backpressure"
+)
+
+type (
+	HealthRequestRateLimiterImpl struct {
+		rateLimiter   *RateLimiterImpl
+		healthSignals backpressure.HealthSignalAggregator
+
+		latencyThreshold    float64
+		errorRatioThreshold float64
+
+		refreshTimer    *time.Timer
+		refreshInterval time.Duration
+
+		// reductionMultiplier will be used to get the new rate limit by multiplying it with the current rate
+		reductionMultiplier float64
+		// increaseStepSize is a ratio of the maxRate that will be added to the current rate when increasing
+		increaseStepSize float64
+
+		maxRate  float64
+		curRate  float64
+		maxBurst int
+		curBurst int
+	}
+)
+
+var _ RequestRateLimiter = (*HealthRequestRateLimiterImpl)(nil)
+
+func NewHealthRateLimiterImpl(
+	healthSignals backpressure.HealthSignalAggregator,
+	refreshInterval time.Duration,
+	reductionMultiplier float64,
+	increaseRatio float64,
+	maxRate float64,
+	maxBurst int,
+) *HealthRequestRateLimiterImpl {
+	return &HealthRequestRateLimiterImpl{
+		rateLimiter:         NewRateLimiter(maxRate, maxBurst),
+		healthSignals:       healthSignals,
+		refreshTimer:        time.NewTimer(refreshInterval),
+		refreshInterval:     refreshInterval,
+		reductionMultiplier: reductionMultiplier,
+		increaseStepSize:    increaseRatio * maxRate,
+		maxRate:             maxRate,
+		curRate:             maxRate,
+		maxBurst:            maxBurst,
+		curBurst:            maxBurst,
+	}
+}
+
+func (rl *HealthRequestRateLimiterImpl) Allow(now time.Time, request Request) bool {
+	rl.maybeRefresh()
+	return rl.rateLimiter.AllowN(now, request.Token)
+}
+
+func (rl *HealthRequestRateLimiterImpl) Reserve(now time.Time, request Request) Reservation {
+	rl.maybeRefresh()
+	return rl.rateLimiter.ReserveN(now, request.Token)
+}
+
+func (rl *HealthRequestRateLimiterImpl) Wait(ctx context.Context, request Request) error {
+	rl.maybeRefresh()
+	return rl.rateLimiter.WaitN(ctx, request.Token)
+}
+
+func (rl *HealthRequestRateLimiterImpl) maybeRefresh() {
+	select {
+	case <-rl.refreshTimer.C:
+		rl.refreshTimer.Reset(rl.refreshInterval)
+		rl.refresh()
+
+	default:
+		// no-op
+	}
+}
+
+func (rl *HealthRequestRateLimiterImpl) refresh() {
```

Comment:
Have you done any simulations of this algorithm? I can see a few issues:

- If the "ideal" rate (to get just under the latency threshold) is less than the max, then this will cause the rate to oscillate a lot (in a sawtooth pattern). You might want some hysteresis or make the response proportional to the excess latency/errors
- If the error rate is high due to an issue not caused by load, after several refresh intervals this will drop the rate to nearly zero. Then even if the problem goes away, it might be hard to get enough successes through to get the error rate below the threshold again, especially if the moving window is near or longer than the refresh interval. Overall it seems quite sensitive to the ratio between the refresh interval and moving window average. You might want to add a min rate also to help with that. And maybe document/warn about how the various parameters should relate